### PR TITLE
Migration flow

### PR DESCRIPTION
**Pull request by tw0po1nt**
Implement the migration flow from an old wallet to a new one

Test steps:
- Pull the code for the [legacy app](https://github.com/nighthawk-apps/zcash-ios-wallet)
- Select 'ECC-Wallet-no-logging' as the target
- Run the app in either the simulator or on a physical device, so long as you use the same device/simulator when testing the migration path.
- Create a wallet in the legacy app (you don't need to let it finish syncing)
- Load up this branch and open the project in Xcode
- Select `secant-mainnet` as the target
- Run the new version of the app on the same device or simulator you used in the previous steps.

Expected result:
You'll see a screen with copy prompting you to either migrate automatically or manually. 

- Choosing 'Continue' should cause the wallet to start syncing with the new internal representation.
- Choosing 'Restore manually' should bring the user to the default welcome screen where they can create a new wallet or manually import an existing one.

In either scenario, the legacy wallet storage will be nuked.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [x] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: modules/Sources/Dependencies/WalletStorage/WalletStorage.swift [hunk 1]
```diff
@@ -82,7 +89,7 @@ public struct WalletStorage {
             throw WalletStorageError.uninitializedWallet
         }
         
-        guard let wallet = try? decode(json: data, as: StoredWallet.self) else {
+        guard let wallet = try decode(json: data, as: StoredWallet.self) else {
```
Comment (tw0po1nt) [inline modules/Sources/Dependencies/WalletStorage/WalletStorage.swift:92]: Reverting it to how it was before, as this change broke some things.

**Review by tw0po1nt — COMMENTED** [review]
Finished self-review

**Comment by pacu** [comment]
I followed the steps and got this error:

```
action:
  RootReducer.Action.initialization(.migrateLegacyWallet)
  RootReducer.State(
-   _alert: nil,
+   _alert: AlertState(
+     title: "Migration failed",
+     message: "An error occurred and we were unable to migrate your wallet automatically. Tap continue to import your wallet manually."
+   ),
    appInitializationState: .needsMigration,
    debugState: RootReducer.DebugState(rescanDialog: nil),
    destinationState: RootReducer.DestinationState(…),
    exportLogsState: ExportLogsReducer.State(…),
    nhHomeState: NHHomeReducer.State(…),
    homeState: HomeReducer.State(…),
    migrateState: MigrateReducer.State(),
    onboardingState: OnboardingFlowReducer.State(…),
    phraseValidationState: RecoveryPhraseValidationFlowReducer.State(…),
    phraseDisplayState: RecoveryPhraseDisplayReducer.State(…),
    sandboxState: SandboxReducer.State(…),
    storedWallet: StoredWallet(…),
    walletConfig: WalletConfig(…),
    welcomeState: WelcomeReducer.State()
  )
  ```
  I'm attaching a video of the result 
  
https://github.com/nighthawk-apps/nighthawk-ios-wallet/assets/484008/cc13e415-501b-4efd-9870-73119475a3b6

File: modules/Sources/Features/Root/RootInitialization.swift [hunk 2]
```diff
@@ -192,6 +195,28 @@ extension RootReducer {
...
                 }
                 return .none
+                
+            case .initialization(.migrateLegacyWallet):
+                do {
+                    let phrase = try walletStorage.exportLegacyPhrase()
+                    let birthday = try walletStorage.exportLegacyBirthday()
+                    
+                    // Delete legacy wallet storage
+                    walletStorage.nukeLegacyWallet()
```
Comment (pacu) [inline modules/Sources/Features/Root/RootInitialization.swift]: shouldn't we attempt to delete the legacy storage after we are sure that they were successfully migrated?

File: secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 3]
```diff
@@ -14,8 +14,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/pointfreeco/combine-schedulers",
       "state" : {
-        "revision" : "ec62f32d21584214a4b27c8cee2b2ad70ab2c38a",
-        "version" : "0.11.0"
+        "revision" : "0625932976b3ae23949f6b816d13bd97f3b40b7c",
+        "version" : "0.10.0"
```
Comment (pacu) [inline secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:18]: possibly unwanted version downgrade

File: secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 4]
```diff
@@ -86,8 +94,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/pointfreeco/swift-clocks",
       "state" : {
-        "revision" : "0fbaebfc013715dab44d715a4d350ba37f297e4d",
-        "version" : "0.4.0"
+        "revision" : "f9acfa1a45f4483fe0f2c434a74e6f68f865d12d",
+        "version" : "0.3.0"
```
Comment (pacu) [inline secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:98]: possibly unwanted version downgrade

File: secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 5]
```diff
@@ -140,8 +139,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/pointfreeco/swift-dependencies",
       "state" : {
-        "revision" : "16fd42ae04c6e7f74a6a86395d04722c641cccee",
-        "version" : "0.6.0"
+        "revision" : "de1a984a71e51f6e488e98ce3652035563eb8acb",
+        "version" : "0.5.1"
```
Comment (pacu) [inline secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:143]: possibly unwanted version downgrade

File: secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 6]
```diff
@@ -248,8 +247,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/pointfreeco/swiftui-navigation",
       "state" : {
-        "revision" : "2aa885e719087ee19df251c08a5980ad3e787f12",
-        "version" : "0.8.0"
+        "revision" : "db81007362f998654239021ca9308a264e59d3e2",
+        "version" : "0.7.2"
```
Comment (pacu) [inline secant.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:251]: possibly unwanted version downgrade

**Review by pacu — CHANGES_REQUESTED** [review]
Unfortunately my manual test failed. :(

Plus code review comments need resolution.

**Comment by pacu** [comment]
More testing feedback.

Apparently is not that I got an error, it's that There's no feedback from the app. 

If the user taps once on the migrate button, the app starts to scan but there's no feedback about it.

If the user force-closes the app and restarts it the app will got to the homescreen.

File: modules/Sources/Features/Root/RootInitialization.swift (hunk 2 above)
Comment (tw0po1nt) [reply modules/Sources/Features/Root/RootInitialization.swift]: Yeah probably!

**Review by pacu — APPROVED** [review]

**Pull request merged**